src/community/features-autopopulate/features-autopopulate-core/src/main/java/org/geoserver/autopopulate/AutopopulateTransactionCallback.java

```diff
@@ -0,0 +1,439 @@
+     * insert/update cases.
+     *
+     * @see TransactionListener#dataStoreChange(TransactionEvent)
+     */
+    @Override
+    public TransactionRequest beforeTransaction(TransactionRequest request) throws WFSException {
+        if (request.getVersion() == null) {
+            return request;
+        }
+
+        List<TransactionElement> newElements = new ArrayList<>();
+        for (TransactionElement element : request.getElements()) {
+            if (element instanceof Insert) {
+                List<SimpleFeature> newFeatures = new ArrayList<>();
+                for (Object of : affectedFeatures(element)) {
+                    if (of instanceof SimpleFeature) {
+                        try {
+                            log.info("Updating feature: " + of);
+                            SimpleFeature transformed = dataStoreChangeInternal((SimpleFeature) of);
+                            newFeatures.add(transformed);
+                        } catch (RuntimeException | IOException e) {
+                            // Do never make the transaction fail due to an
+                            // AutopopulateTransactionCallback error.
+                            // Yell on the logs though
+                            log.log(
+                                    Level.WARNING,
+                                    "Error pre computing the transaction's affected attributes",
+                                    e);
+                            newFeatures.add((SimpleFeature) of);
+                        }
+                    }
+                }
+                ((Insert) element).setFeatures(newFeatures);
+                newElements.add(element);
+            } else if (element instanceof Update) {
+                FeatureTypeInfo featureTypeInfo =
+                        getFeatureTypeInfo(new NameImpl(element.getTypeName()));
+                try {
+                    SimpleFeatureCollection features = getTransactionFeatures((Update) element);
+                    SimpleFeatureIterator featuresIterator = features.features();
+                    while (featuresIterator.hasNext()) {
+                        SimpleFeature feature = featuresIterator.next();
+                        log.info("Updating feature: " + feature);
+                        SimpleFeature transformed = dataStoreChangeInternal(feature);
+                        List<Property> properties = ((Update) element).getUpdateProperties();
+                        for (org.geotools.api.feature.Property p : transformed.getProperties()) {
+                            if (properties.stream()
+                                    .anyMatch(
+                                            prop ->
+                                                    prop.getName()
+                                                            .getLocalPart()
+                                                            .equals(p.getName().getLocalPart()))) {
```

Comment:
This bit is used twice and is fighing with the max number of columns in the formatter. Maybe have a simple support method, ``match(p1, p2)`` method that compares the local names instead?

Reply:
Done